src/ray/pubsub/subscriber.cc

```diff
@@ -208,16 +222,38 @@ void Subscriber::HandleLongPollingResponse(const rpc::Address &publisher_address
     RAY_LOG(DEBUG) << "A worker is dead. subscription_failure_callback will be invoked. "
                       "Publisher id: "
                    << publisher_id;
+
+    // First get all failure callbacks from the channel.
+    std::queue<std::shared_ptr<SubscriptionFailureCallback>> failure_callbacks;
     for (const auto &channel_it : channels_) {
       channel_it.second->HandlePublisherFailure(publisher_address);
     }
+    // Empty the command queue because we cannot send commands anymore.
+    commands_.erase(publisher_id);
   } else {
     // Otherwise, iterate on the reply and pass published messages to channels.
+    std::queue<std::shared_ptr<SubscriptionCallback>> subscription_callbacks;
     for (int i = 0; i < reply.pub_messages_size(); i++) {
       const auto &msg = reply.pub_messages(i);
       const auto channel_type = msg.channel_type();
-      Channel(channel_type)->HandlePublishedMessage(publisher_address, msg);
+      subscription_callbacks.push(
+          Channel(channel_type)->GetCallbackForPubMessage(publisher_address, msg));
+    }
+    // We need to unlock here because the subscription callback could call Subscriber APIs
+    // which needs to hold a lock. Note that we don't allow to call subscriber APIs inside
+    // the failure callback right now, so we don't need to do the same thing for that.
+    mutex_.Unlock();
```

Comment:
I feel bad about this one. Could you do some effort to avoid explicit unlock? I don't even know why it's locked here. :(

Reply:
Actually the reasoning is inside the comment. There are various fields that need to be protected in this method. And we should unlock here otherwise subscribe callback will deadlock if it uses any Subscriber APIs.

Reply:
This needs to be protected because this method is called within the client callback, which runs on a different io service. So the method should be locked.

Reply:
Also it is okay to unlock for all code block inside (it just calls a callback). Alternative approach is posting them to io service, but this can add pretty big overhead

Reply:
The third approach is to use reentrantlock lock (but I didn't really like this approach)

Reply:
Okay. So there are 4 approaches. Lmk which one you prefer the most; 

- The current (explicit unlock)
- pass io service to subscriber. This can make things very ugly if the client pool doesn't use the same io service as passed io service (so I don't prefer it).
- re-entarant lock
- Lock twice (before this block + after this block) explictly

Reply:
I think we talked about post the job to io services in core worker (not the subscriber). Did I miss anything? The reason this is bad is that it's hard to understand even with the comment.

Reply:
Let me make an issue and follow up later in a separate PR. I am still not sure if it is a good behavior to make this asynchronous (it is much harder to use /reason for API user perspective whereas the current approach is just a little complex for the module developer, but this sort of unlocking is not too uncommon pattern in our code).

Reply:
I'm open to all options but strong against the lock here.

Reply:
Let me try understanding what you are against here actually. Is it just that it is hard to understand why we need to unlock here (even with comments)? So using re-entarant lock can be a possible option for you?

Reply:
With lock here makes things complicated, but I'm ok with giving it a go (not strong opinion).
but we shouldn't have things like:
```
// in one place
{
   lock
}
// in another place
{
  unlock
}
```

Previously, we have 
```
{
   lock_guard(lock)
}
```
I'm fine with this. I remember last time when we have lock and unlock explicitly, I also mentioned that it's hard to understand the code. The thing here is that we can do something to avoid this one, not the case that we have to use lock/unlock explicitly.

This usually is because the thread modeling has some issues which make it hard to implement.

Reply:
So, there's no clear alternative that can ensure the same behavior except re-entarant lock? I think making the api caller use io service.post is worse than this because it requires the api caller to think about threading model + implementation detail.